### PR TITLE
wrap client http response body in buffered input stream to support mark/reset

### PR DESCRIPTION
**Pull request by noffke**
## Description
When using `LogbookClientHttpRequestInterceptor` on spring resttemplate and the the underlying HTTP library (e.g. apache httpclient) returns an input stream that doesn't support mark/reset the response can't be logged by logbook.

## Motivation and Context
Fixes the second exception mentioned in #963

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java [hunk 1]
```diff
@@ -62,6 +62,7 @@ public State without() {
         @Override
         public State buffer(final ClientHttpResponse response) throws IOException {
             InputStream responseBodyStream = response.getBody();
+            responseBodyStream.mark(Integer.MAX_VALUE);
```
Comment (whiskeysierra) [inline logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java:65]: Why is that needed?

File: logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
+package org.zalando.logbook.spring;
```
Comment (whiskeysierra) [inline logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java:1]: It would be nice to have a more realistic test which verifies that a stream can be read multiple times.

File: logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java (hunk 1 above)
Comment (noffke) [reply logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java:65]: Doesn't a stream need to be marked before it can be reset?

File: logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java (hunk 2 above)
Comment (noffke) [reply logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java:1]: Do you have a suggestion for a more realistic test? The Spring Mock Server always returns a byte array input stream which doesn't exhibit the original problem.

File: logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java (hunk 1 above)
Comment (whiskeysierra) [reply logbook-spring/src/main/java/org/zalando/logbook/spring/RemoteResponse.java:65]: You're probably right. Interesting that this never caused any issue before, since we had `reset` in there for a while.

File: logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java (hunk 2 above)
Comment (noffke) [reply logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java:1]: Ok, I've adapted the test to do real network requests instead of mock requests. This way, if the `BufferingClientHttpResponseWrapper` is not used, the original "mark/reset not supported" exception will be thrown.

Comment (whiskeysierra) [reply logbook-spring/src/test/java/org/zalando/logbook/spring/BufferingClientHttpResponseWrapperTest.java:1]: > This way, if the `BufferingClientHttpResponseWrapper` is not used, the original "mark/reset not supported" exception will be thrown.

Perfect! Exactly the kind of test I like for this kind problem.

**Comment by whiskeysierra** [comment]
Mmh, the spring4 tests are failing because of lack of coverage. Any clue why that is?

**Comment by noffke** [comment]
I have unfortunately no clue why the spring4 tests fail. I originally also wasn't able to run them locally. When commenting out the `logbook-servlet` module in the root pom, I'm able to progress the build to the `logbook-spring` module. I was then not able to make it work as a spring boot test so I've changed the test to use wiremock instead. After that, I was able to successfully run "spring4" profile locally.

**Comment by noffke** [comment]
@whiskeysierra did I miss anything here for the PR to be approved?

**Comment by whiskeysierra** [comment]
Sorry for the delay, I was on vacation.

Coming back to this I finally remembered why we added reset to begin with. It's meant to be a more efficient alternative to buffering. I feel that it should be changed to the following:

- check if mark/reset is supported (using `markSupported()`)
- if it's supported, then mark it in the beginning, then reset afterwards
- if it's not suppoirted, wrap the stream to enable buffering

**Comment by noffke** [comment]
Thanks for the clarification. I'll see what I can do but I'm rather busy at the moment so it might take a moment to follow up, sorry.

**Comment by DaniloVeljovic** [comment]
@noffke Any follow ups here? We would be interested in merging this PR, if you would resolve @whiskeysierra remarks

**Comment by noffke** [comment]
@DaniloVeljovic sorry, completely forgot about this, too much work :-) Does the PR look acceptable now?

**Comment by DaniloVeljovic** [comment]
LGTM 👍🏼
